README.md
# floweRy
R package to manage Celery task queues via Flower's
[REST API](https://flower.readthedocs.io/en/latest/api.html).

## Install

```R
install.packages("remotes")
remotes::install_github("nuno-agostinho/floweRy")
```

## Setup

By default, flower runs in http://localhost:5555. If this is not the case,
change the default URL:

```R
options(flowerURL="http://flower:5555")
```

## Run and manage tasks

Tasks can be run in two ways:
- Apply: run task and wait for result
- Async-apply: run task in background (do not wait for result)

```R
# Add two numbers using tasks.add
taskApply <- taskApply("tasks.add", 3, 4)
taskAsync <- taskAsyncApply("tasks.add", 20, 53)

taskResult(taskAsync) # get task result
taskInfo(taskAsync)   # get task info (including state)

# List all tasks
taskList()
```

|uuid                                 |name      |state   |received            |sent |started             |rejected |succeeded           |failed |retried |revoked |args     |kwargs |eta |expires | retries|result |exception |timestamp           |   runtime|traceback |exchange |routing_key | clock|client |root                                 |root_id                              |parent |parent_id |children |worker              |
|:------------------------------------|:---------|:-------|:-------------------|:----|:-------------------|:--------|:-------------------|:------|:-------|:-------|:--------|:------|:---|:-------|-------:|:------|:---------|:-------------------|---------:|:---------|:--------|:-----------|-----:|:------|:------------------------------------|:------------------------------------|:------|:---------|:--------|:-------------------|
|a7a67c47-d373-4e77-9445-9679a7f1c9cf |tasks.add |SUCCESS |2021-09-07 18:38:34 |NA   |2021-09-07 18:38:34 |NA       |2021-09-07 18:38:34 |NA     |NA      |NA      |[20, 53] |{}     |NA  |NA      |       0|73     |NA        |2021-09-07 18:38:34 | 0.0015033|NA        |NA       |NA          |    16|NA     |a7a67c47-d373-4e77-9445-9679a7f1c9cf |a7a67c47-d373-4e77-9445-9679a7f1c9cf |NA     |NA        |NULL     |celery@2d5665fb19f4 |
|e694adc8-3d4d-43f0-a9d6-c6553139dbc9 |tasks.add |SUCCESS |2021-09-07 18:38:34 |NA   |2021-09-07 18:38:34 |NA       |2021-09-07 18:38:34 |NA     |NA      |NA      |[3, 4]   |{}     |NA  |NA      |       0|7      |NA        |2021-09-07 18:38:34 | 0.0049620|NA        |NA       |NA          |    13|NA     |e694adc8-3d4d-43f0-a9d6-c6553139dbc9 |e694adc8-3d4d-43f0-a9d6-c6553139dbc9 |NA     |NA        |NULL     |celery@2d5665fb19f4 |

```R
# Revoke task
task <- taskAsyncApply("tasks.add", 3, 4)
taskRevoke(task)
```

## Manage workers

```R
# List all workers
workerList()
worker <- names(workerList())[[1]]

# Manage worker's pool
workerPoolGrow(worker, 5)
workerPoolShrink(worker, 3)
workerPoolAutoscale(worker, 3, 7)
workerPoolRestart(worker)

# Shutdown worker
workerShutdown(worker)
```


NEWS.md
# floweRy 0.1.0

* Added a `NEWS.md` file to track changes to the package.


R/utils.R
getFlowerURL <- function() {
    getOption("flowerURL", "http://localhost:5555")
}

#' @importFrom httr status_code stop_for_status
simplifyAPIerror <- function(res, errors) {
    msg <- errors[[as.character(status_code(res))]]
    if (!is.null(msg)) {
        stop(msg)
    } else {
        stop_for_status(res)
    }
}

#' @importFrom httr GET content stop_for_status
getInfo <- function(type, ..., url=getFlowerURL(), errors=NULL) {
    res <- GET(url=url, path=file.path("api", type), query=list(...))
    simplifyAPIerror(res, errors)
    return(content(res))
}

#' @importFrom httr POST stop_for_status content
postAction <- function(type, action, id, ..., url=getFlowerURL(), errors=NULL) {
    url <- file.path(url, "api", type, action, id)
    res <- POST(url, query=list(...))

    simplifyAPIerror(res, errors)
    return(content(res))
}

#' @importFrom httr POST content
runTask <- function(type="apply", task=NULL,
                    args=NULL, kwargs=NULL, options=NULL, url=getFlowerURL(),
                    errors=NULL) {
    path <- file.path("api/task", type, task)

    convert2list <- function (x) if (!is.null(x) && !is.list(x)) list(x) else x
    args    <- convert2list(args)
    kwargs  <- convert2list(kwargs)
    options <- convert2list(options)

    body <- list(args=args, kwargs=kwargs, options=options)
    res  <- POST(url=url, path=path, body=body, encode="json")
    simplifyAPIerror(res, errors)
    return(content(res))
}


R/taskApply.R
#' Execute a task
#'
#' Tasks can be executed using:
#' - `taskApply()`: execute a task and wait for result (better for quick tasks)
#' - `taskAsyncApply()`: execute a task and immediately return task info without
#' result (better for long-running tasks)
#' - `taskSend()`: execute a task (doesn't require task sources)
#'
#' @param task Character: task to run
#' @param args List of arguments
#' @param kwargs Dictionary of arguments
#'
#' @return Task information:
#'  - Status
#'  - Identifier
#'  - Result (in case of `taskApply()` only)
#' @export
#'
#' @examples
#' taskApply("tasks.add")
taskApply <- function(task, args=NULL, kwargs=NULL, url=getFlowerURL()) {
    errors <- list("404"="unknown task")
    runTask(type="apply", task=task, args=args, kwargs=kwargs, url=url,
            errors=errors)
}

#' @rdname taskApply
#'
#' @param options Dictionary of *apply_async* keyword arguments
#' @export
#'
#' @examples
#' taskAsyncApply("tasks.add", args=list(a=1, b=2))
taskAsyncApply <- function(task, args=NULL, kwargs=NULL, options=NULL,
                           url=getFlowerURL()) {
    errors <- list("404"="unknown task")
    runTask(type="async-apply", task=task,
            args=args, kwargs=kwargs, options=options, url=url, errors=errors)
}

#' @rdname taskApply
#' @export
#'
#' @examples
#' taskSend("tasks.add", args=list(a=1, b=2))
taskSend <- function(task, args=NULL, kwargs=NULL, url=getFlowerURL()) {
    errors <- list("404"="unknown task")
    runTask(type="send-task", task=task, args=args, kwargs=kwargs, url=url,
            errors=errors)
}


R/taskActions.R
#' Task operations
#'
#' @param id Character: task identifier
#' @inheritParams workerList
#'
#' @return Flower response
#' @export
#'
#' @examples
#' task <- rownames(taskList())[[1]]
#' taskAbort(task)
taskAbort <- function(id, url=getFlowerURL()) {
    errors <- list("503"="result backedn is not configured")
    postAction(url=url, "task", "abort", id, errors=errors)
}

#' @rdname taskAbort
#' @param terminate Boolean: terminate task if running?
#' @param signal Character: signal to send when terminating task
#'
#' @export
#' @examples
#' taskRevoke(task)
taskRevoke <- function(id, terminate=TRUE, signal="SIGTERM",
                       url=getFlowerURL()) {
    postAction(url=url, "task", "revoke", id, terminate=terminate,
               signal=signal)
}

#' @rdname taskAbort
#' @param workername Character: worker name
#' @param softLimit Integer: soft limit
#' @param hardLimit Integer: hard limit
#'
#' @examples
#' worker <- names(workerList())[[1]]
#' taskTimeout(worker, "tasks.add", softLimit=60)
taskTimeout <- function(workername, task, softLimit=30, hardLimit=100,
                        url=getFlowerURL()) {
    errors <- list("404"="unknown task/worker")
    postAction(url=url, "task", "timeout", task, workername=workername,
               softLimit=softLimit, hardLimit=hardLimit, errors=errors)
}

#' @rdname taskAbort
#' @param ratelimit Integer: rate limit
#'
#' @export
#' @examples
#' taskRateLimit(worker, "tasks.add", 200)
taskRateLimit <- function(workername, task, ratelimit, url=getFlowerURL()) {
    errors <- list("404"="unknown task/worker")
    postAction(url=url, "task", "rate-limit", task,
               ratelimit=ratelimit, workername=workername, errors=errors)
}


celery/tasks.py
#!/usr/bin/env python 
import os, subprocess, time
from datetime import datetime

# celery configuration
from celery import Celery
os.environ.setdefault('C_FORCE_ROOT', 'true')
app = Celery("tasks",
             broker=os.environ.get('CELERY_BROKER_URL', 'redis://redis'),
             backend=os.environ.get('CELERY_RESULT_BACKEND', 'redis://redis'))
app.conf.CELERY_WORKER_SEND_TASK_EVENTS = True

def createDir(dir):
    os.mkdir(dir)
    os.chdir(dir)

def rmEmptyDir(dir):
    if len( os.listdir(dir) ) == 0:
        os.rmdir(dir)

@app.task
def add(a, b):
    return a + b

@app.task(bind=True)
def R(self, cmd):
    # Use task ID to create new folder to save task data
    taskID = self.request.id
    createDir(taskID)

    # Run R process
    cmd = "R -e '" + cmd + "'"
    subprocess.run(cmd, shell=True)

    # Go back from folder and remove task folder if empty
    os.chdir("..")
    rmEmptyDir(taskID)
    return "hello"

if __name__ == "__main__":
    app.start()


R/workerActions.R
#' Worker operations
#'
#' @inheritParams workerList
#'
#' @return Flower response
#' @export
#'
#' @examples
#' worker <- names(workerList())[[1]]
#' workerShutdown(worker)
workerShutdown <- function(workername, url=getFlowerURL()) {
    errors <- list("404"="unknown worker")
    postAction(url=url, "worker", "shutdown", workername, errors=errors)
}

#' @rdname workerShutdown
#' @export
#' @examples
#' workerPoolRestart(worker)
workerPoolRestart <- function(workername, url=getFlowerURL()) {
    errors <- list(
        "403"="pool restart is not enabled (see CELERYD_POOL_RESTARTS)",
        "404"="unknown worker")
    postAction(url=url, "worker/pool", "restart", workername, errors=errors)
}

#' @rdname workerShutdown
#'
#' @param n Integer: number of pool processes to grow/shrink
#'
#' @export
#' @examples
#' workerPoolGrow(worker, 4)
workerPoolGrow <- function(workername, n=1, url=getFlowerURL()) {
    errors <- list("403"="failed to grow", "404"="unknown worker")
    postAction(url=url, "worker/pool", "grow", workername, n=n, errors=errors)
}

#' @rdname workerShutdown
#'
#' @export
#' @examples
#' workerPoolShrink(worker, 4)
workerPoolShrink <- function(workername, n=1, url=getFlowerURL()) {
    errors <- list("403"="failed to shrink", "404"="unknown worker")
    postAction(url=url, "worker/pool", "shrink", workername, n=n, errors=errors)
}

#' @rdname workerShutdown
#'
#' @param min Integer: minimum number of pool processes
#' @param max Integer: maximum number of pool processes
#'
#' @export
#' @examples
#' workerPoolAutoscale(worker, 4)
workerPoolAutoscale <- function(workername, min, max, url=getFlowerURL()) {
    errors <- list(
        "403"="autoscaling is not enabled (see CELERYD_AUTOSCALER)",
        "404"="unknown worker")
    postAction(url=url, "worker/pool", "autoscale", workername,
               min=min, max=max, errors=errors)
}

#' @rdname workerShutdown
#'
#' @param queue Character: queue name
#'
#' @export
#' @examples
#' workerQueueAddConsumer(worker, "sample-queue")
workerQueueAddConsumer <- function(workername, queue, url=getFlowerURL()) {
    errors <- list("403"="failed to add consumer", "404"="unknown worker")
    postAction(url=url, "worker/queue", "add-consumer", workername,
               queue=queue, errors=errors)
}

#' @rdname workerShutdown
#'
#' @export
#' @examples
#' workerQueueCancelConsumer(worker, "sample-queue")
workerQueueCancelConsumer <- function(workername, queue, url=getFlowerURL()) {
    errors <- list("403"="failed to cancel consumer", "404"="unknown worker")
    postAction(url=url, "worker/queue", "cancel-consumer", workername,
               queue=queue, errors=errors)
}


R/getters.R
#' List workers
#'
#' @param refresh Boolean: update list of workers
#' @param workername Character: filter info by worker's name
#' @param status Boolean: only get information on worker status
#' @param url Character: Flower URL
#'
#' @return List of workers containing worker information
#' @export
#'
#' @examples
#' workerList()
#' workerList(refresh=TRUE)
#' workerList(status=TRUE) # only print worker status
#' workerList(workername="celery@d2b3d2b64077")
workerList <- function(refresh=FALSE, workername=NULL, status=FALSE,
                       url=getFlowerURL()) {
    getInfo(type="workers", refresh=refresh, workername=workername,
            status=status, url=url)
}

#' List information of tasks
#'
#' @param limit Integer: maximum number of tasks
#' @param offset Integer: skip first N tasks
#' @param sortBy Character: sort tasks by attribute (`name`, `state`,
#' `received`, `started`)
#' @param workername Character: filter task by workername
#' @param taskname Character: filter tasks by taskname
#' @param state Character: filter tasks by state
#' @param receivedStart Character: filter tasks by received date (must be
#' greater than) format %Y-%m-%d %H:%M
#' @param receivedEnd Character: filter tasks by received date (must be less
#' than) format %Y-%m-%d %H:%M
#' @param table Boolean: format response as table? If not, a list is returned
#' @inheritParams workerList
#'
#' @return Data frame or list of task information
#'
#' @export
#' @examples
#' taskList()
taskList <- function(limit=NULL, offset=NULL,
                     sort_by=c("name", "state", "received", "started"),
                     workername=NULL, taskname=NULL, state=NULL,
                     received_start=NULL, received_end=NULL, table=TRUE,
                     url=getFlowerURL()) {
    sort_by <- match.arg(sort_by)
    res <- getInfo(type="tasks", limit=limit, offset=offset, sort_by=sort_by,
                   workername=workername, taskname=taskname, state=state,
                   received_start=received_start, received_end=received_end,
                   url=url)

    timestamps <- c("received", "started", "succeeded", "timestamp", "revoked")
    convertTime <- function(x) as.POSIXct(x, origin="1970-01-01")

    for (task in names(res)) {
        # Replace NULL with NA
        nulls <- vapply(res[[task]], is.null, logical(1))
        res[[task]][nulls] <- NA

        # Replace Unix timestamp with formatted time
        res[[task]][timestamps] <- lapply(res[[task]][timestamps], convertTime)
    }

    # Create table
    if (table) {
        cols <- unique(unlist(lapply(res, names)))
        df   <- data.frame(matrix(ncol=length(cols), nrow=length(res),
                                  dimnames=list(names(res), cols)))
        for (col in cols) {
            values <- sapply(res, "[[", col)
            if (col %in% timestamps) values <- convertTime(values)
            df[[col]] <- values
        }
        res <- df
    }
    return(res)
}

#' List (seen) task types
#'
#' @inheritParams workerList
#'
#' @return List of task types
#' @export
#'
#' @examples
#' taskTypesList()
taskTypesList <- function(url=getFlowerURL()) {
    getInfo(type="task/types", url=url)
}

#' List length of all active queues
#'
#' @inheritParams workerList
#'
#' @return List of active queues
#' @export
#'
#' @examples
#' queueLength()
queueLength <- function(url=getFlowerURL()) {
    erros <- list("503"="result backend is not configured")
    getInfo(type="queues/length", url=url, errors=errors)
}

#' Get a task info or results
#'
#' @param id Character: task identifier
#' @inheritParams workerList
#'
#' @return Information or results on tasks
#' @export
#'
#' @examples
#' task <- rownames(taskList())[[1]]
#' taskInfo(task)
taskInfo <- function(id, url=getFlowerURL()) {
    errors <- list("404"="unknown task")
    getInfo(type=file.path("task", "info", id), url=url, errors=errors)
}

#' @rdname taskInfo
#' @param timeout Integer: seconds to wait before operation times out
#'
#' @export
#'
#' @examples
#' taskResult(task)
taskResult <- function(id, timeout=NULL, url=getFlowerURL()) {
    errors <- list("503"="result backedn is not configured")
    getInfo(type=file.path("task", "result", id), timeout=timeout, url=url,
            errors=errors)
}
